### PR TITLE
Extend exportable code

### PR DESCRIPTION
**Pull request by kprokofi**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by jaegukhyun — DISMISSED** [review]
Just for the check, Will otx export {exported model} replace otx demo in OTX1.x?

**Comment by kprokofi** [comment]
> Just for the check, Will otx export {exported model} replace otx demo in OTX1.x?

Yes. `otx deploy` and `otx demo` are replaced with `otx export`. Demo is now stand alone code, that used without OTX package.

**Comment by kprokofi** [comment]
@eunwoosh , could you take a look at HPO issue? Did you face this problem?
image

**Comment by eunwoosh** [comment]
> @eunwoosh , could you take a look at HPO issue? Did you face this problem? image

Could you check again? HPO test is set as 'xfailed', because it doesn't work in develop branch due to refactoring. It is being fixed currently.

File: src/otx/core/exporter/base.py [hunk 1]
```diff
@@ -150,25 +163,45 @@ def to_exportable_code(
...
             precision (OTXExportPrecisionType, optional): precision of the exported model's weights
+            path_to_already_exported_model (PathLike, optional): path to the already
+                exported model to add it in exportable code
 
         Returns:
             Path: path to the exported model.
         """
-        work_dir = Path(demo.__file__).parent
+        work_demo_dir = Path(demo.__file__).parent
+        training_extensions_dir = Path(otx.__file__).parents[2]
```
Comment (eunwoosh) [inline src/otx/core/exporter/base.py]: I think it doesn't work if otx is installed as non-editable mode. Could you copy LICENSE to demo directory and use it? You can refer #3255

**Review by vinnamkim — COMMENTED** [review]
Is this PR about the ticket no. 135467? I still don't know why don't we just drop `OTXExportFormatType.EXPORTABLE_CODE` and introduce a boolean input argument such as `package_model_into_exportable_code: bool`. If we add this, the number of argument increases, but this PR too. In addition, the usage introduced by this PR seems more complicated to me.

Is there any special reason that we must keep `OTXExportFormatType.EXPORTABLE_CODE`?

**Comment by kprokofi** [comment]
> Is this PR about the ticket no. 135467? I still don't know why don't we just drop `OTXExportFormatType.EXPORTABLE_CODE` and introduce a boolean input argument such as `package_model_into_exportable_code: bool`. If we add this, the number of argument increases, but this PR too. In addition, the usage introduced by this PR seems more complicated to me.
> 
> Is there any special reason that we must keep `OTXExportFormatType.EXPORTABLE_CODE`?

I don't see the difference in complexity.
If we remove OTXExportFormatType.EXPORTABLE_CODE, everything will remain same except that we need to pass additional "--save-as-zip" or whatever. It gets even more complicated, since we need to pass additional parameters to methods of the model class

Now, user just need to: otx export ... --export_type EXPORTABLE_CODE
If we remove it: otx export ... --save-as-zip

Is it make sense to delete additional export type?

**Comment by vinnamkim** [comment]
> I don't see the difference in complexity. If we remove OTXExportFormatType.EXPORTABLE_CODE, everything will remain same except that we need to pass additional "--save-as-zip" or whatever. It gets even more complicated, since we need to pass additional parameters to methods of the model class
> 
> Now, user just need to: otx export ... --export_type EXPORTABLE_CODE If we remove it: otx export ... --save-as-zip
> 
> Is it make sense to delete additional export type?

Ticket no. 135467 is about packaging an optimized model (INT8), not exported model (FP16 or FP32) into an exportable code format (zip file). Following this PR, I imagine that the CLI workflow will be 
```console
otx optimize <some-path-1>
otx export --import <some-path-1> --export <some-path-2> -- export_type EXPORTABLE_CODE
```
I abbreviate a Python API case but it looks similar.

**Comment by kprokofi** [comment]
> > I don't see the difference in complexity. If we remove OTXExportFormatType.EXPORTABLE_CODE, everything will remain same except that we need to pass additional "--save-as-zip" or whatever. It gets even more complicated, since we need to pass additional parameters to methods of the model class
> > Now, user just need to: otx export ... --export_type EXPORTABLE_CODE If we remove it: otx export ... --save-as-zip
> > Is it make sense to delete additional export type?
> 
> Ticket no. 135467 is about packaging an optimized model (INT8), not exported model (FP16 or FP32) into an exportable code format (zip file). Following this PR, I imagine that the CLI workflow will be
> 
> ```
> otx optimize <some-path-1>
> otx export --import <some-path-1> --export <some-path-2> -- export_type EXPORTABLE_CODE
> ```
> 
> I abbreviate a Python API case but it looks similar.

Optimized or exported models - doesn't matter. It is all about OpenVINO models

```
otx optimize ...
otx export ... --checkpoint path/to/.xml
```

If you pass .xml to export, you don't even need to change export_type, it will automatically proceed with exportable code

**Comment by vinnamkim** [comment]
> Optimized or exported models - doesn't matter. It is all about OpenVINO models
> 
> otx optimize ... otx export --checkpoint path/to/.xml
> 
> If you pass .xml to export, you don't even need to change export_type

I'm asking why don't we abbreviate it to one step such as
```console
otx optimize --package-to-exportable-code
```
, and Python API too.

**Comment by kprokofi** [comment]
> > Optimized or exported models - doesn't matter. It is all about OpenVINO models
> > otx optimize ... otx export --checkpoint path/to/.xml
> > If you pass .xml to export, you don't even need to change export_type
> 
> I'm asking why don't we abbreviate it to one step such as
> 
> ```
> otx optimize --package-to-exportable-code
> ```
> 
> , and Python API too.



> > Optimized or exported models - doesn't matter. It is all about OpenVINO models
> > otx optimize ... otx export --checkpoint path/to/.xml
> > If you pass .xml to export, you don't even need to change export_type
> 
> I'm asking why don't we abbreviate it to one step such as
> 
> ```
> otx optimize --package-to-exportable-code
> ```
> 
> , and Python API too.

Ok, I understand what you mean. I think we can add this functionality. I can add additional argument to `engine.optimize` and run `engine.export` after optimization if user wants to export models with demo

File: src/otx/core/exporter/base.py (hunk 1 above)
Comment (kprokofi) [reply src/otx/core/exporter/base.py]: Ok, let's add LICENSE to demo folder

**Comment by kprokofi** [comment]
I added possibility to export demo during optimization + removed additional parameter in export with path to xml model. I found a better way

**Review by vinnamkim — DISMISSED** [review]

**Review by sovrasov — APPROVED** [review]

**Review by eunwoosh — APPROVED** [review]

**Pull request merged**